redash/models.py

```diff
@@ -365,6 +386,7 @@ class User(TimestampMixin, db.Model, BelongsToOrgMixin, UserMixin, PermissionsCh
     __table_args__ = (db.Index('users_org_id_email', 'org_id', 'email', unique=True),)
 
     def __init__(self, *args, **kwargs):
+        kwargs["email"] = kwargs.get("email", None).lower()
```

Comment:
Isn't this handled by the `EmailType`?

Reply:
I added this, to handle the case where the email as input is returned directly to the API response.
Without it, it triggers the following failure:

```
======================================================================
FAIL: test_creates_user_case_insensitive_email (tests.handlers.test_users.TestUserListResourcePost)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/tests/handlers/test_users.py", line 37, in test_creates_user_case_insensitive_email
    self.assertEqual(rv.json['email'], 'user@example.com')
AssertionError: u'User@Example.com' != 'user@example.com'
```

When a [User is created here](https://github.com/getredash/redash/blob/0c404fa6024569651110d21f27a41eab10f29419/redash/handlers/users.py#L31-L34), the input response is sent directly back, despite the value being stored in the DB with the lowercased value.

A reproduction via shell:

```python
>>> from redash import models
>>> org = models.Organization.query.first()
>>> user = models.User(org=org, name="Fred", email="FRED@flintstone.org")
>>> models.db.session.add(user)
>>> models.db.session.commit()
>>> print(user.email)
FRED@flintstone.org

# New shell session
>>> from redash import models
>>> user = models.User.query.filter_by(email="FRED@flintstone.org").one()  # Note: this is where the custom comparison comes into play
>>> print(user.email)
fred@flintstone.org
```

So either there's something to be done about resetting the session post-commit, or re-fetching the value from the DB. Using `kwargs` to ensure that when instantiating a `User()`, the email always gets lowered.

If it's acceptable that the response to a user being created contain the original case, then I'll remove it, and change the test case assertion - I'm not clear how the response is used yet, so wanted to ensure that whatever is returned is something the caller can depend on.

Reply:
Reading a little more, we could add a `refresh()` call, which expires the sqlalchemy object state and immediately performs a SELECT.

Which is preferable?

- Ensuring each time a `User()` object is created, a call to `lower()` happens in Python-land, and from then on the session is internally consistent
- When pushing a value to the DB, we trigger a second query to refresh the session object's state from the DB

Reply:
👍 sounds good. Let's keep it the way you implemented to avoid an extra query. One thing I noticed though: `kwargs.get('email', None)` might return `None`, in which case calling `.lower()` on it will fail.

We should probably call `.lower()` only if we get back a value. Maybe something like:

```python
if kwargs.get('email') is not None:
    kwargs['email'] = kwargs['email'].lower()
```

(`.get` returns `None` by default)